### PR TITLE
Enable nullability in more small line count files

### PR DESCRIPTION
**Pull request by elachlan**
Related: #1107

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/8649)

### CODE REVIEW

**Comment by elachlan** [comment]
CC: @gpetrou

File: src/System.Windows.Forms/src/System/Windows/Forms/TabControl.ControlCollection.cs [hunk 1]
```diff
@@ -19,11 +17,11 @@ public ControlCollection(TabControl owner) : base(owner)
                 _owner = owner;
             }
 
-            public override void Add(Control value)
+            public override void Add(Control? value)
             {
-                if (!(value is TabPage))
+                if (value is not TabPage)
```
Comment (halgab) [inline src/System.Windows.Forms/src/System/Windows/Forms/TabControl.ControlCollection.cs:22]:
```suggestion
                if (value is not TabPage tabPage)
```
and delete line 27? The same pattern can be found on line 70.

File: src/System.Windows.Forms/src/System/Windows/Forms/PropertyGridInternal/MergePropertyDescriptor.MultiMergeCollection.cs [hunk 2]
```diff
@@ -12,7 +10,7 @@ internal partial class MergePropertyDescriptor
     {
         private class MultiMergeCollection : ICollection
         {
-            private object[] _items;
+            private object[]? _items;
```
Comment (lonitra) [inline src/System.Windows.Forms/src/System/Windows/Forms/PropertyGridInternal/MergePropertyDescriptor.MultiMergeCollection.cs:13]: Should this be `object?[]?`

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewSelectedRowCollection.cs [hunk 3]
```diff
@@ -28,16 +26,16 @@ void IList.Clear()
...
+        bool IList.Contains(object? value) => ((IList)_items).Contains(value);
 
-        int IList.IndexOf(object value) => ((IList)_items).IndexOf(value);
+        int IList.IndexOf(object? value) => ((IList)_items).IndexOf(value);
 
-        void IList.Insert(int index, object value)
+        void IList.Insert(int index, object? value)
         {
             throw new NotSupportedException(SR.DataGridView_ReadOnlyCollection);
         }
```
Comment (lonitra) [inline src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewSelectedRowCollection.cs:36]: We can clean these up a little more by doing the following in the appropriate places
```suggestion
        void IList.Insert(int index, object? value) => throw new NotSupportedException(SR.DataGridView_ReadOnlyCollection);
```

**Review by lonitra — COMMENTED** [review]
There are some `TabControlControlCollection` tests that are expecting `NullReferenceException` and should be fixed to expect an `ArgumentException` if we are not expecting to be able to add `null`. If we need to make that change this might also be documentation breaking since the exception type has changed it seems.

File: src/System.Windows.Forms/src/System/Windows/Forms/TabControl.ControlCollection.cs [hunk 4]
```diff
@@ -19,13 +19,13 @@ public ControlCollection(TabControl owner) : base(owner)
 
             public override void Add(Control? value)
             {
-                if (value is not TabPage)
+                value.OrThrowIfNull(nameof(value));
+
+                if (value is not TabPage tabPage)
                 {
                     throw new ArgumentException(string.Format(SR.TabControlInvalidTabPageType, value?.GetType().Name));
```
Comment (halgab) [inline src/System.Windows.Forms/src/System/Windows/Forms/TabControl.ControlCollection.cs:26]:
```suggestion
                    throw new ArgumentException(string.Format(SR.TabControlInvalidTabPageType, value.GetType().Name));
```
you already checked `value` for `null` on line 22

**Review by lonitra — COMMENTED** [review]
👍

**Comment by dreddy-work** [comment]
Captured failed error here: https://github.com/dotnet/winforms/issues/8598#issuecomment-1435102622

File: src/System.Windows.Forms/src/System/Windows/Forms/TabControl.ControlCollection.cs (hunk 4 above)
Comment (elachlan) [reply src/System.Windows.Forms/src/System/Windows/Forms/TabControl.ControlCollection.cs]: `OrThrowIfNull` does not have the annotations which indicate the input is not null if returned. So replaced with `ArgumentNullException.ThrowIfNull`.

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewSelectedColumnCollection.cs [hunk 5]
```diff
@@ -69,13 +67,7 @@ internal DataGridViewSelectedColumnCollection()
...
         protected override ArrayList List => ArrayList.Adapter(_items);
 
-        public DataGridViewColumn this[int index]
-        {
-            get
-            {
-                return _items[index];
-            }
-        }
+        public DataGridViewColumn this[int index] => _items[index];
```
Comment (dreddy-work) [inline src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewSelectedColumnCollection.cs:70]: Q: if `_items[index]` is nullable on line 48, why not here? Build successful suggests this is not nullable?

Comment (elachlan) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewSelectedColumnCollection.cs:70]: `IList` contract for `this` requires nullability. Where as the `List<DataGridViewColumn>` we have at the top of the class isn't nullable. So in the instance you pointed out, its not nullable.

**Pull request merged**